### ISSUE TITLE
DS4 (Dualshock 4) Does not work on PC over Bluetooth, but does over USB

### ISSUE THREAD

**Issue by Jack54guythecoder** [hrydgard/ppsspp#19429]
### Game or games this happens in

NONE, but I cant play anything because of this

### What area of the game / PPSSPP

Step 1: Have a Standard issue DS4.
Step 2: With it paired, press the PS Button.
Step 3: The DS4 does not work.

I did try Touchscreen Test, but nothing happened as well.
Also, in gamecontrollerdb.txt, theres mention of various PS4 controllers.

### What should happen

You should be able to play games with a PS4 Controller with bluetooth (wireless)

### Platform

Windows

### Mobile device model or graphics card (GPU)

NVIDIA GeForce RTX 3050 Laptop GPU running on driver 546.80.0.0

### PPSSPP version affected

v1.17.1-1009

### Last working version

I do not know

### Graphics backend (3D API)

Vulkan

### Checklist

- [X] Test in the [latest git build](https://www.ppsspp.org/devbuilds) in case it's already fixed.
- [x] [Search for other reports](https://github.com/hrydgard/ppsspp/search?q=my+issue&type=issues) of the same issue.
- [X] Try resetting settings or older versions and include if the issue is related.
- [X] Try without any cheats and without loading any save states.
- [X] Include logs or screenshots of issue.

**Comment by anr2me** [comment]
Does other emulators have similar issue? or only happened on PPSSPP?

**Comment by Jack54guythecoder** [comment]
> Does other emulators have similar issue? or only happened on PPSSPP?

Only PPSSPP.

**Comment by Jack54guythecoder** [comment]
@hrydgard  I was the one who talked to you on Discord about this.
Any idea about whats going on?

**Comment by hrydgard** [comment]
According to https://www.playstation.com/en-us/support/hardware/ps4-pair-dualshock-4-wireless-with-pc-or-mac/ , DS4 is only supported via USB on PC.

However, since you say it works on other emulators, I assume there's some way around it, so I'm going to have to investigate what those other emulators are doing to make it work...

**Comment by NABN00B** [comment]
DS4 controllers connected through Bluetooth definitely work natively with PPSSPP. The problem lies elsewhere.

There are two ways to connect a DS4 to a computer, wired and wireless.
When you switch the method of connection
- from USB to BT: the controller will be somewhat recognized, although as a different device, and will act weird (eg reverse analog controls in menus).
- from BT to USB: the controller will no longer be recognized.
Both scenarios can be fixed by either restarting PPSSPP, or switching the connection method back to the one PPSSPP initially detected.

I would like to emphasize that this was always the case as long as I've used PPSSPP.

In practice this means that if you initially connect via BT and then the battery runs out and need to connect via cable, the controller will no longer work with PPSSPP. The foolproof method is connecting initially via BT and only then plugging the USB into the DS4. This way PPSSPP will never lose detection because the battery will be charging.

**Comment by Jack54guythecoder** [comment]
@hrydgard  It work over Bluetooth with other emulators, RPCS3 for example has its dedicated DS4 handler.
They  DO NOT use SteamInput AT ALL.
Also, I do not need to use Steam or DS4Windows for the other emulators, as they work natively with my gamepad.

As for the support via usb...... that is just to expose some DS4 features such as the little speaker on it.

@NABN00B  This is a bit of a doozy, for other emus (at least for me), it works no matter how you connect it or the order in which you do it.
I have windows 10 and dont have to do this for other emulator software.

UPDATE ON SITUATION> Now it does not work over USB at all, neither bluetooth.
Not even a log console message.

**Comment by hrydgard** [comment]
I really have no clue what's going in with it working intermittently. I've written proper XInput and DirectInput code that tries to read input from any connected gamepads, that SHOULD work, at least for USB.

It's a shame that specialized code apparently needs to be written for this controller to work. Just makes the lives of software developers more painful....

But I guess at some point I'll have to get around to looking at what RPCS3 does then.

Or you can use an XBox controller which will just work.

**Comment by NABN00B** [comment]
> @NABN00B This is a bit of a doozy, for other emus (at least for me), it works no matter how you connect it or the order in which you do it. I have windows 10 and dont have to do this for other emulator software.

It also works for me on PCSX2 and Project64 using Octomino's SDL plugin. I can test melonDS and BizHawk later if necessary.

**Comment by NineKain** [comment]
PPSSPP wont recognize my controller over BT with Steam open (and steam input enabled) its the only emulator that does this

**Comment by SandyVoidez** [comment]
if i may, i would like to add  a little bit of my own experience of PPSSPP and DS4:

so i have following setup :

- two DS4 controllers,  one is an old model, another one is newer (i think theese are gen1 and gen2, G2 is the one that has a color bleed line on the top of the touchpad)  
- rescent version of PPSSPP (1.17 i think) through standalone exe in portable install mode
- both controllers are conntected to pc with win 10 through bluetooth. 
- several emulators (pcsx2, retroarch etc)
- steam. 
- no software which may modify or cloak the controller is installed (DS4Windows for example)

it works like this:
_steam isn't running (or started in the background), ppsspp is stared from the exe:_

1. ppsspp  will **not** recognize G1 controller  (there may be a way when both controllers are unpaired, removed from windows and only G1 is connected, but i'm not so sure on this)
2. both controllers are recognized by different emulators (retroarch, pcsx2) and in steam
3. windows wont recognize  G1 input at all (controller is connected, but in test there is no response)
4. pcsx2 allows for controllers hot-swap, basically when one is out of battery i could turn another controller on and continue playing without additional setup (i will check later if i did bound both controllers to p1 gamepad) 

so if i use just ppsspp i may use only G2 controller,
next, i've checked that g1 and g2 are recognized by steam and added ppsspp to my steam library as an external game. 
_When ppsspp is started through steam library:_
this way steam handles the input from controller and masks them as an x360 gamepads. 
with additional controller mapping in the emulator it has allowed me to use both controllers in the ppsspp

**Comment by NineKain** [comment]
I think most emus use SDL, just fyi if it could be implemented

**Comment by Jack54guythecoder** [comment]
> I think most emus use SDL, just fyi if it could be implemented

Yeah, PCSX2 and Ducstation use SDL, meaning that I can use my DS4 natively.

**Comment by hrydgard** [comment]
Maybe we should give up our custom code and leave it to SDL on Windows as well. I'm not really happy about adding a dependency though...

**Comment by NineKain** [comment]
I understand, but its such a good improvement, and most other emus do this

**Comment by Jack54guythecoder** [comment]
Update: Now it is working CONSISTENTLY over USB.

Would like to be able to use thru Bluetooth tho.

**Comment by NABN00B** [comment]
> Maybe we should give up our custom code and leave it to SDL on Windows as well. I'm not really happy about adding a dependency though...

Perhaps the switch to SDL would have other benefits in the long run eg when new consoles come out in the future or regarding other currently unsupported controllers. I've had a friend report a problem with the Switch controller back in 2023.

![Image](https://github.com/user-attachments/assets/fb8da1cb-974c-4117-a4dc-9631af4d6796)

**Comment by hrydgard** [comment]
I'm gonna close this as a duplicate of https://github.com/hrydgard/ppsspp/issues/20473, we have too many issues about this. A fix is coming soon (gonna use the HID API directly for these controllers).
